Feature/saemi

shinpanda opened the pull request:
## 요구사항

#2
- [x]  주문 목록 페이지에는 주문에 대한 모든 정보를 표 형태로 확인할 수 있어야 합니다.
- [x]  주문 목록은 페이지네이션이 구현되어야 합니다(한 페이지에 50건의 주문이 보여야 합니다)
- [x]  데이터 중에서 오늘의 거래건만 보여지도록 해주세요
  - 여기서 오늘은 “2023-03-08” 일을 의미합니다.
#3 
- [x]  기본 정렬은 ID 기준 오름차순으로 구현해주세요
- [x]  표에서 주문번호, 거래일 & 거래시간 버튼을 누르면 각각 내림차순 정렬이 되도록 해주세요
- [x]  주문 처리 상태에 따라 filtering 기능을 구현해주세요
- [x]  고객이름을 검색할 수 있도록 해주세요
- [ ]  서버에 들어온 주문을 5초마다 최신화 해주세요 - 테스트 진행 못함

#4 
OrderList 컴포넌트에 대해서는 테스트 코드를 작성하였으나, 데이터를 호출하고 날짜별로 필터링 하는 부분은 작성하지 못하였습니다.

## 구현 사항 설명
[구현 기능 설명]
1. 주어진 데이터를 이용하여 주문 목록 페이지를 구현해주세요
- 주어진 데이터 받아오기
    mock data는 public 폴더의 data에 넣고 axios로 get을 해 받아올 수 있도록 함.
- 주문 목록 페이지에는 주문에 대한 모든 정보를 표 형태로 확인할 수 있어야 합니다.
    chakra ui 를 사용하여 UI를 그림.

- 주문 목록은 페이지네이션이 구현되어야 합니다(한 페이지에 50건의 주문이 보여야 합니다)
    받은 데이터를 50건씩 잘라 보이게 하고, 표 아래에 paging 버튼을 두어, 버튼을 눌렀을 때,
    해당 인덱스의 데이터가 보일 수 있도록 구현

- 데이터 중에서 오늘의 거래건만 보여지도록 해주세요
    - 여기서 오늘은 **“2023-03-08”**일을 의미합니다.
    처음에는 날짜가 하나의 날짜만 아니고 여러 날짜이기 때문에 최신 날짜를 default 값으로 하여,
    구현을 진행하였으나, 요구사항이 특정 날짜가 '2023-03-08'로 고정된 것 같아, 하드코딩으로 대체
    추후 변동 가능성이 있으므로 해당 코드는 변경이 필요할 것으로 보임

#3 
- 정렬/검색/필터 의 기능은 useSearchParam을 이용하여 구현하였습니다.
- 정렬의 경우, 각각의 버튼을 누르면 내림차순으로 정렬하게 하였으나,
주문번호, 거래일& 거래시간을 같이 눌렀을 때에도 정렬이 되도록 쿼리를 append하도록 하였고,
주문번호를 먼저 정렬 후 거래일 거래시간을 같이 내림차순으로 정렬하게 하였습니다.
- 필터(주문 처리 상태)의 경우 처리/비처리 상태로 나눠 적용되어야 하므로 set으로 이전 검색값을 덮도록 하였습니다.
- 고객 이름 검색의 경우에도 set으로 이전 검색값을 덮도록 하였습니다.
- 서버에 들어온 주문을 5초마다 최신화 해주세요 -> queryClient에 staletime을 5초로 설정하였는데 테스트는 진행하지 못하였습니다.

#4 
mock Data를 ts파일에 넣고 export 하여, orderList 컴포넌트에서 수행하는 작업에 대하여 컴포넌트를 작성하였습니다.

## 보완할 점
- 검색 기능 관련
검색 기능을 OrderList에 같이 추가하여 날짜별로 고객의 데이터를 보이게 했으나,
고객 이름으로 검색 시, 전체 날짜의 데이터를 보이게 하는 것이 좀 더 좋지 않을까 합니다.
현재 존재하는 데이터로 날짜를 뽑아 select로 선택하게 하고 있으나,
전체라는 옵션을 추가하여, 고객 이름 검색 시에는 날짜에 상관없이 검색이 가능하도록 하고,
검색을 하고 나서도 거래일이 특정 날짜로 선택이 되면 해당 날짜, 해당 고객 데이터가 보이도록 해놔야 좀 더 검색이 편리할 것으로 생각됩니다.

- 테스트 코드를 작성하다 보니, 데이터 가공하는 부분을 따로 뒀어야 했나라는 생각을 많이 들게 했습니다.
저의 경우, OrderPage에서 데이터를 가져온 뒤, 날짜별로 우선 잘라 선택한 날짜의 데이터만 orderList에 넣어주고 있었는데요.
이렇게 나눴던 이유는 정렬이나, 주문 처리 상태, 검색, 페이지네이션 모두 이미 표로 그려진 데이터 내에서 작동하는 것이라 orderList에 포함되는 게 맞다고 생각하였고, OrderPage에서는 어떠한 데이터가 넘어갈 것인지를 결정하고 orderList에 던져주는 역할을 주었는데, 뭔가 역할이 너무 많았나 라는 생각을 하였습니다.

- 생각해보니 더 쪼개서 테스트 코드를 짰어야 했는데 하는 생각이 듭니다.
페이징, 정렬, 검색 들도 컴포넌트로 분리 하고 따로 테스트 코드를 짠다면, 이후 기능이 변경되거나 추가 될 때, 어디서 오류가 발생하는지 찾기 쉬울테니까요.

[관련 스크린샷 첨부]
![image](https://user-images.githubusercontent.com/30384031/226899283-e94ec599-de74-4160-8fe6-3d7bd05c68f2.png)
![image](https://user-images.githubusercontent.com/30384031/226899327-9a7fbd9e-8fe3-40dd-8513-bacf6f6681fa.png)
1. 페이지 이동 시 (3페이지로 이동)
![image](https://user-images.githubusercontent.com/30384031/226899415-3066e4aa-cf0c-4c98-8118-4fe7c48a475d.png)
2. 주문 번호 클릭 시 
![image](https://user-images.githubusercontent.com/30384031/226899703-549fc79b-3df9-4d1c-a8df-946180eed281.png)
3. 거래일 & 거래시간 클릭 시
![image](https://user-images.githubusercontent.com/30384031/226899780-8a7734a2-a183-440e-9c6f-b05b626919f2.png)
4. 필터링
  4-1. 처리
  ![image](https://user-images.githubusercontent.com/30384031/226899865-fdf269ad-98d1-4bf2-be8b-94374ef4a6a2.png)
  4-2. 비처리
  ![image](https://user-images.githubusercontent.com/30384031/226899913-1a26cafc-6238-423f-af72-a5c4e115a6e1.png)
5. OrderList에 대한 테스트
![image](https://user-images.githubusercontent.com/30384031/226906186-10bad421-b4df-4492-804e-f518ab0e4d4a.png)

shinpanda committed:
feature : 주어진 데이터를 이용하여 주문 목록 페이지를 구현
1. 주어진 데이터를 이용하여 주문 목록 페이지를 구현해주세요
- 주어진 데이터 받아오기
    mock data는 public 폴더의 data에 넣고 axios로 get을 해 받아올 수 있도록 함.
- 주문 목록 페이지에는 주문에 대한 모든 정보를 표 형태로 확인할 수 있어야 합니다.
    chakra ui 를 사용해서 표를 그리도록 하였다.

- 주문 목록은 페이지네이션이 구현되어야 합니다(한 페이지에 50건의 주문이 보여야 합니다)
    받은 데이터중 50건만 보이게 하고, 표 아래에 paging 버튼을 두어, 버튼을 눌렀을 때,
    해당 인덱스의 데이터가 보일 수 있도록 구현하였다.

- 데이터 중에서 오늘의 거래건만 보여지도록 해주세요
    - 여기서 오늘은 **“2023-03-08”**일을 의미합니다.
    처음에는 날짜가 하나의 날짜만 아니고 여러 날짜이기 때문에 최신 날짜를 default 값으로 하여,
    구현을 진행하였으나, 요구사항이 특정 날짜가 '2023-03-08'로 고정된 것 같아, 하드코딩으로 대체하였다.
    추후 변동 가능성이 있으므로 해당 코드는 변경이 필요하다.
[#2]

shinpanda committed:
refactor : 거래일 및 주문 건 표시를 OrderPage로 변경

OrderPage에서 처음 데이터를 가져오고 낣짜별 데이터를 OrderList에 넘기면
OrderList는 해당되는 데이터들만 목록으로 보여주도록 하는 것이 컴포넌트 역할을 줄일 듯 하여 이동

shinpanda committed:
refactor : 거래일 선택박스를 OrderPage로 변경
OrderPage에서 처음 데이터를 가져오고 낣짜별 데이터를 OrderList에 넘기면
OrderList는 해당되는 데이터들만 목록으로 보여주도록 하는 것이 컴포넌트 역할을 줄일 듯 하여 이동
주문 건수는 리스트에 뿌려질 데이터이므로 OrderList에서 이동하지 않음
[#2]

shinpanda committed:
feature : 정렬 기능 추가
OrderList 컴포넌트에 책임이 많아지고 있어 분리가 필요
[#3]

shinpanda committed:
feature :  필터링 기능 추가
주문 처리 상태를 버튼으로 선택할 수 있도록 구현
[#3]

shinpanda committed:
feature : 검색 기능 추가
검색 기능을 OrderList에 같이 추가하여 날짜별로 고객의 데이터를 보이게 했으나,
고객 이름으로 검색 시, 전체 날짜의 데이터를 보이게 하는 것이 좀 더 사용성이 있을 것 같음.
현재 존재하는 데이터로 날짜를 뽑아 select로 선택하게 하고 있으나,
전체라는 옵션을 추가하여, 고객 이름 검색 시에는 날짜에 상관없이 검색이 가능하도록 하고,
검색을 하고 나서도 거래일이 특정 날짜로 선택이 되면 해당 날짜, 해당 고객 데이터가 보이도록 하는 것이 좋을 것 같음.
[#3]

shinpanda committed:
refactor : 함수명 오타 변경

shinpanda committed:
refactor: 함수 분리 및 컴포넌트 분리 진행
주문 목록 안에 검색, 페이징 부분을 따로 컴포넌트로 분리.
필터 부분도 진행이 필요함

shinpanda committed:
fix : 없는 날짜로 데이터 조회 시 오류 발생
없는 날짜로 데이터를 조회하더라도 해당 일에 주문이 없다고 문구가 떠야 함

shinpanda committed:
test : OrderList 에 대한 테스트 코드 추가
eslint에서 테스트코드 작성 시 빨간 줄을 보이고 있어
eslint-plugin-jest 라이브러리 설치 및 eslintrc에서 jest import 추가하였습니다.
user-event setup() 함수가 작성되지 않아 검색하였더니 최신 버전으로 다운받으면 된다고 하여
최신 버전으로 변경하였더니 테스트가 정상적으로 동작하였습니다.

[#4]

shinpanda committed:
style : mock 데이터 폴더 변경
_mock_안에 data 에 데이터를 이동

shinpanda committed:
fix: .eslintrc 에 jest import 부분 수정

shinpanda committed:
fix: .eslintrc jest 플러그인 import를 eslint-plugin-jest으로 변경

shinpanda committed:
fix: paging 관련 1페이지 일 경우 컴포넌트가 안그려져 다시 작성

Seo0H commented on src/_test_/OrderList.test.js:
```diff
@@ -0,0 +1,147 @@
+
+describe('<OrderList />', () => {
+  describe('주문 목록이 없을 경우', () => {
+    render(<OrderList orderLists={[]} />, { wrapper: BrowserRouter });
```
아 이런식으로 그냥 컴포넌트 단위로 테스트 해봐야 겠네요,,,, 다 랜더링되고 테스트해야한다고 잘못 생각했네요 😢 코드에 적용 해봐야겠어요

Seo0H commented on src/pages/error/NotFound.tsx:
```diff
@@ -0,0 +1,5 @@
+const NotFound = () => {
+  return <div>404 Not Found</div>;
+};
+
+export default NotFound;
```
에러  페이지 만들어 두신거 좋은 거 같아요!

Seo0H reviewed (commented):
코드 잘 봤습니다 새미님!

yoosion030 commented on src/index.tsx:
```diff
@@ -1,16 +1,25 @@
-const root = ReactDOM.createRoot(
-  document.getElementById('root') as HTMLElement
-);
+const queryClient = new QueryClient({
+  defaultOptions: {
+    queries: {
+      staleTime: 5000,
+    },
+  },
+});
```
react-query를 사용하면 데이터 refetch를 이렇게 간단하게 구현할 수 있는거였군요.. 당장 공부해봐야겠네요!

yoosion030 reviewed (commented):
새미님! 짧은 시간에도 이렇게나 많은 테스트 케이스를 작성하시다니.. 정말 대단하신 것 같아요. 필터링이나 날짜 선택 select, 깔끔한 UI가 인상깊은 것 같아요! 수고하셨습니다 ^ . ^

yoosion030 commented on src/components/order/OrderList.tsx:
```diff
@@ -0,0 +1,153 @@
+  return filter !== null ? datas.filter(order => '' + order.status === filter) : datas;
+};
+
+const searchDatasByCustomerName = (datas: Order[], search: string | null) => {
+  return search !== null ? datas.filter(order => order.customer_name === search) : datas;
+};
```
새미님
https://pre-onboarding-9th-4-10-git-fea-a517b6-wanted-onboarding-10team.vercel.app/
제가 여기서 고객 이름 검색을 했는데 데이터가 안떠서 당황했는데, 풀네임으로 검색되게 하셨더라구요. UX적으로 생각해보았을 때 
 includes를 사용해서 검색하는 방식은 어떠신가요?

shinpanda replied:
안그래도 그 부분에 대해 고민이 좀 있긴 했어요 ㅠ 역시 그 편이 좋겠죠? ㅎㅎ...

yoosion030 replied:
앜ㅋㅋㅋㅋㅋㅋㅋ 검색할 때마다 렌더링이 일어나는 것도 아니고 그러니 그 편이 좋을 것 같습니다~!

Ryong-E commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,77 @@
+import { getOrders } from 'utils/api/order';
+
+const OrderPage = () => {
+  const { isLoading, data } = useQuery<Order[]>('orders', getOrders);
```
5초마다 리패칭하는 부분이 빠진거같아요

shinpanda replied:
queryClient 에 staleTime을 5초로 설정하면 캐싱한다고 알고 있어서 useQuery에다가는 설정하지 않았는데, 한번 해당 내용을 확인해 봐야겠네요 감사합니다 ㅎ

Ryong-E commented on src/pages/OrderPage.tsx:
```diff
@@ -0,0 +1,77 @@
+  // TODO: '2023-03-08' 하드코딩 제거
+  const selectedDate = searchParams.get('date') ?? '2023-03-08';
+
+  const orderListsBytransactionDate = data
+    ? data.reduce((prev, curr) => {
+        const dateTime = curr.transaction_time.split(' ');
+        prev[dateTime[0]] = prev[dateTime[0]] || [];
+        prev[dateTime[0]].push(curr);
+        return prev;
+      }, init)
+    : {};
+
+  const existTransactionDates = Object.keys(orderListsBytransactionDate).sort();
+  const orderListsByDate = existTransactionDates.includes(selectedDate)
+    ? orderListsBytransactionDate[selectedDate]
+    : [];
+
+  const changeDate = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setSearchParams(searchParams => {
+      searchParams.delete('page');
+      searchParams.set('date', event.target.value);
+      return searchParams;
+    });
+  };
```
분리를 하면 좋을거같아요

shinpanda replied:
맞아요.. 분리하려고 했는데 거기까지 가지 못했네요 큐ㅠㅠ 끝나면 개인적으로 리팩토링 해보려구요 큐ㅠ

Ryong-E commented on src/components/order/OrderList.tsx:
```diff
@@ -0,0 +1,153 @@
+  });
+
+  const totalOrders = viewDatas ? viewDatas.length - 1 : 0;
+  const maxIndex =
```
마지막 페이지를 구하는것같습니다
viewDatas가 필터링을 거치고 마지막인것같은데
viewDatas의 길이를 한 페이지당 50페이지이니 Math.ceil(viewDatas / 50)만 해도 될것같아요
50건이 되지않아도 페이지는 있어야하니 올림으로 해주어야합니다

Ryong-E commented on src/components/order/OrderList.tsx:
```diff
@@ -0,0 +1,153 @@
+    Math.floor(totalOrders / pageSize) - 1 > 0 ? Math.floor(totalOrders / pageSize) - 1 : 0;
+  const pagingIndex = parseInt(searchParams.get('page') ?? '1') - 1;
+
+  const setSortKey = (key: string) => {
```
주문번호 정렬과, 거래시간을 동시에 정렬할 수는 없을 것 같아요
현재 무슨 정렬을 사용하고있는지 알려주는게 좋을거같습니다

shinpanda replied:
음 동시에 한다고 한다면 거래시간을 먼저 정렬하고 주문 번호를 같이 하는 방향은 있을 것 같아요.
거래시간이 겹치는 경우들은 있을 수 있으니까요.. 음 그래도 해당 부분은 좀 더 고민해 보겠습니다 ㅎ

shinpanda committed:
feature : 고객 이름 검색 시 성이나 이름만 입력해도 검색이 되도록 수정
이름을 모두 작성하지 않았을 때에도 검색이 된다면 좀 더 ux적으로 편리할 것 같다는
의견을 적용하여, includes로 일부만 작성해도 검색이 가능하도록 함.
또한, 주어진 데이터가 영문으로 되어 있어 비교 시, 소문자로 변경하여 검색을 진행하도록 함.

Ryong-E commented on src/components/order/OrderList.tsx:
```diff
@@ -0,0 +1,153 @@
+  };
+
+  const setFiltering = (status: boolean) => {
+    setSearchParams(searchParams => {
```
setSearchParams를 이렇게 바꾸면
콜백함수에 인자값으로 현재 쿼리 파라미터가 들어오는건가요?
위쪽에 useSearchParams에서 받아오는 이름과 콜백 프로퍼티 이름이 같아서 헷갈리네요

shinpanda replied:
네 맞아요 setSearchParams 안에 함수로 위와 같이 하면 현재 값(변경 전의 값)이 파라메터로 들어가게 됩니다. 위쪽의 searchParams랑 파라메터의 값은 동일해서 따로 수정을 하진 않았어요.

Ryong-E commented on src/_test_/OrderList.test.js:
```diff
@@ -0,0 +1,147 @@
+  });
+
+  describe('필터', () => {
+    const mockSearchParams = jest.fn();
```
리턴하는 값이 없는데
useSearchParams 사용하는곳에서 문제가 일어나진 않나요?

shinpanda replied:
네네 이렇게 했을 때에도 아래 테스트 사용 시 문제가 일어나진 않았어요.
아래 테스트를 not을 빼기도 하고, 다시 넣으면서 테스트가 정상적인지는 확인했습니다.
다만 query 자체에 대해 테스트코드를 완성하지는 못해서 그 부분을 추가하면 좋을 것 같아요

shinpanda committed:
fix : 정렬 수정
정렬 부분 거래시간이 먼저 이루어진 후 주문 번호가 정렬되도록 수정

shinpanda committed:
fix: refectch 설정을 하여, 5초마다 데이터 최신화

shinpanda committed:
feature : 데이터에 따라 헤더와 본문이 자동으로 그려지도록 컴포넌트 분리 및 커스텀 훅 분리
- 데이터에 따라 헤더와 본문이 자동으로 그려지도록 컴포넌트 분리
    - 데이터가 추가되고 삭제 될 때, 그려지는 순서라든가, 변경할 곳이 최소한으로 줄일 수 있도록 변경하였습니다.
- 커스텀 훅 분리
    - 쿼리를 사용하는 방식 외에도 변경될 가능성이 있는 부분으로 따로 관리할 수 있도록 변경이 필요하다 판단하였습니다.
